src/Rules/PHPUnit/DataProviderHelper.php

```diff
@@ -0,0 +1,111 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Rules\PHPUnit;
+
+use PHPStan\Analyser\Scope;
+use PHPStan\PhpDoc\ResolvedPhpDocBlock;
+use PHPStan\PhpDocParser\Ast\PhpDoc\PhpDocTagNode;
+use PHPStan\Reflection\MissingMethodFromReflectionException;
+use PHPStan\Rules\RuleError;
+use PHPStan\Rules\RuleErrorBuilder;
+use function array_merge;
+use function preg_match;
+use function sprintf;
+use function trim;
+
+class DataProviderHelper
+{
+
+	/**
+	 * @return array<PhpDocTagNode>
+	 */
+	public function getDataProviderAnnotations(?ResolvedPhpDocBlock $phpDoc): array
+	{
+		if ($phpDoc === null) {
+			return [];
+		}
+
+		$phpDocNodes = $phpDoc->getPhpDocNodes();
+
+		$annotations = [];
+
+		foreach ($phpDocNodes as $docNode) {
+			$annotations = array_merge(
+				$annotations,
+				$docNode->getTagsByName('@dataProvider')
+			);
+		}
+
+		return $annotations;
+	}
+
+	/**
+	 * @return RuleError[] errors
+	 */
+	public function processDataProvider(
+		Scope $scope,
+		PhpDocTagNode $phpDocTag
+	): array
+	{
+		$dataProviderName = $this->getDataProviderName($phpDocTag);
+		if ($dataProviderName === null) {
+			// Missing name is already handled in NoMissingSpaceInMethodAnnotationRule
+			return [];
+		}
+
+		$classReflection = $scope->getClassReflection();
+		if ($classReflection === null) {
+			// Should not happen
+			return [];
+		}
+
+		try {
+			$dataProviderMethodReflection = $classReflection->getMethod($dataProviderName, $scope);
```

Comment:
Needs to use `getNativeMethod`. I doubt that PHPUnit executes `__call`.

Reply:
changed